### ISSUE TITLE
Feature Request: WAF functionality leveraging OWASP CRS, implemented and enabled by default

### ISSUE THREAD

**Issue by dune73** [memorysafety/river#8]
Creating this feature request was recommended by @drcaramelsyrup at

https://github.com/cloudflare/pingora/issues/31#issuecomment-2000798482

OWASP CRS currently runs on the following WAF engines:
* OWASP ModSecurity v2
* OWASP ModSecurity v3
* OWASP Coraza

Commercial integrations are done via custom implementations of the rule language. This includes the Cloudflare setup.

If a new open source Reverse Proxy is created, then giving it WAF functionality based on the de facto standard rule set from the beginning would be useful.

**Comment by mcpherrinm** [comment]
I think WAF type functionality will be a great use of the wasm extensibility planned for River: Rule sets could be compiled into wasm to reject requests.

**Comment by fzipi** [comment]
This should basically alter the proposed design. Adding a new `Filters` between `Listeners` and `Connectors` should do the trick. Then a WAF is just another filter in the chain.

**Comment by dune73** [comment]
Thank you @jamesmunns. I do not really have spare time to contribute here in my volunteer time capacity, but if you have any questions about CRS or especially input on how to provide a successful integration, then please get in touch.

**Comment by jamesmunns** [comment]
Will do, thanks!

**Comment by fzipi** [comment]
Got it now, I see that it is totally under that stage. Awesome. Ping me if you need anything.

**Comment by gmreads** [comment]
**TLDR;** Is achieving ~50k req/sec of 4-5KB payload for CRS Ruleset configured with paranoia-level/1 possible ?

I have been deep diving into WAFs and CRS  trying to build a WAF, we are also an open source organization. 

Most implementations seems to be quite slow as per [Coraza Benchmarks](https://coraza.io/docs/reference/benchmarks/) highest throughput seems to be a mere 1000 req/sec.

I have gone through most of cloudflare's blogs and especially [John's talk.](https://www.youtube.com/watch?v=nlt4XKhucS4) . , but can't seem to find how to actually achieve extremely high throughput.

From my understanding it looks like eventually regex matching is the bottleneck. Using RE2 regex engine is even more slower at least in Java. 

I was trying to benchmark the [RuleId 932330 (RCE-Regexes)](https://github.com/coreruleset/coreruleset/blob/41cfd07d51a97bfd5b71d9f7faa69e966078cf9d/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf#L631). 

For a request payload of 4-5KB , the speed is just 1000 records in 12 seconds. 

Would love to discuss on WAFs and CRS performance, happy to help implement this for River. 


PS: I am very new to rust and performance engineering in general, please excuse my stupidity 😄 .

**Comment by dune73** [comment]
For the record: Performance is an issue. And it does boil down to regex processing and in practical use the ability to handle rule exclusions at run time (that's how you fight false positives).

Whatever Coraza is claiming, the fastest CRS implementation I have seen is ModSec 2.9.x on Apache 2.4.x. That's probably the benchmark.

Performance optimizations work via rule exclusions too, but surprisingly there is a frequent situation where the rule exclusion is more resource intense then executing the rule itself due to sub-optimal implementation in ModSec. We do not currently know about Coraza in this regard AFAICT.

**Comment by SonNgo2211** [comment]
@jamesmunns Does River have any plans for further development?
